gpcontrib/arenadata_toolkit/src/file_hook.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * File hooks to track events.
+ */
+
+#include "file_hook.h"
+
+#include "postgres.h"
+#include "storage/smgr.h"
+#include "access/xact.h"
+#include "catalog/namespace.h"
+#include "utils/lsyscache.h"
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#include "catalog/objectaccess.h"
+
+#include "tf_shmem.h"
+#include "arenadata_toolkit_guc.h"
+#include "drops_track.h"
+
+static file_create_hook_type next_file_create_hook = NULL;
+static file_extend_hook_type next_file_extend_hook = NULL;
+static file_truncate_hook_type next_file_truncate_hook = NULL;
+static file_unlink_hook_type next_file_unlink_hook = NULL;
+
+static bloom_t * non_committed_bloom = NULL;
+static Oid	non_committed_dbid = InvalidOid;
+
+static bool
+is_file_node_trackable(RelFileNodeBackend * rnode)
+{
+	return !(rnode->node.dbNode == InvalidOid);
+}
+
+static void
+file_node_set(RelFileNodeBackend * rnode)
+{
+	if (!is_file_node_trackable(rnode))
+		return;
+
+	bloom_set_set(&tf_shared_state->bloom_set, rnode->node.dbNode, rnode->node.relNode);
+}
+
+/* 'create' events stored in local bloom and merged only on commit, when changes are already in catalog */
+static void
+xact_end_create_callback(XactEvent event, void *arg)
+{
+	if (event != XACT_EVENT_COMMIT && event != XACT_EVENT_ABORT)
+		return;
+
+	elog(DEBUG1, "xact_end_create_callback");
+
+	if (event == XACT_EVENT_COMMIT)
+		bloom_set_merge(&tf_shared_state->bloom_set, non_committed_dbid, non_committed_bloom);
+
+	pfree(non_committed_bloom);
+	non_committed_bloom = NULL;
+	non_committed_dbid = InvalidOid;
+}
+
+static void
+hook_create(RelFileNodeBackend rnode)
+{
+	uint64_t	hash;
+
+	if (next_file_create_hook)
+		next_file_create_hook(rnode);
+
+	if (!is_file_node_trackable(&rnode))
+		return;
+
+	if (!non_committed_bloom)
+	{
+		non_committed_bloom =
+			MemoryContextAlloc(TopMemoryContext, FULL_BLOOM_SIZE(bloom_size));
+		bloom_init(bloom_size, non_committed_bloom);
+		non_committed_dbid = rnode.node.dbNode;
+		RegisterXactCallbackOnce(xact_end_create_callback, NULL);
+	}
+
+	elog(DEBUG1, "hook_create: %d %d %d %d",
+	rnode.backend, rnode.node.dbNode, rnode.node.spcNode, rnode.node.relNode);
+
+	hash = bloom_set_calc_hash(&rnode.node.relNode, sizeof(rnode.node.relNode));
+	bloom_set(non_committed_bloom, hash);
+
+}
+
+static void
+hook_extend(RelFileNodeBackend rnode)
+{
+	if (next_file_extend_hook)
+		next_file_extend_hook(rnode);
+
+	elog(DEBUG1, "hook_extend: %d %d %d %d",
+	rnode.backend, rnode.node.dbNode, rnode.node.spcNode, rnode.node.relNode);
```

Comment:
Please indent arguments here and in similar places.

Reply:
Now it's better. more or less